woocommerce/woocommerce
# Simplify Customer history calculation

### Changes proposed in this Pull Request:

There are some indications that the Order Attribution feature is reducing performance on some stores. The two most likely culprits in the feature seem to be (a) injecting checkout blocks dependency unnecessarily (addressed in https://github.com/woocommerce/woocommerce/pull/43857) and the process of calculating a customer's order history, used during checkout and when viewing the Order Edit page in the dashboard.

This PR addresses possible extra processing power in the customer's order history in two ways:

1. ~Removes order history data requirement during checkout; not retrieving/calculating this information should relieve any extra processing time during checkout added by this requirement.~
    **Note:** this was cherry picked into WooCommerce 8.6 with #44359 and #44395
2. Simplifies customer order history processing in the Order Edit form:
    1. Uses the Customers Query to retrieve customer data rather than calculating it manually to match that data.
    2. This data is calculated/updated asynchronously so it's possible that a the customer history for a customer who placed an order in the past X seconds will be out of date.
    3. This data is only present when **WooCommerce > Settings > Advanced > Features > Enable WooCommerce Analytics** is selected. If that's not the case, then the Customer History box will not be visible on the order edit page. 

**Notes**: 
- https://github.com/woocommerce/woocommerce/pull/44318#discussion_r1477975192, this data will be added to the `wc_customer_lookup` table for easier reference.
- The Customer history will likely be visible in the future, but with a message to enable WC Analytics to see the history.

### Screenshots

See that the numbers match up still:

_WooCommerce Analytics Customer report_
![image](https://github.com/woocommerce/woocommerce/assets/228780/005e6c59-68df-4875-aaa6-d3114e58e13b)

_Customer history metabox_
![image](https://github.com/woocommerce/woocommerce/assets/228780/9e72f1fb-affb-4a90-aab3-784166688255)


### How to test the changes in this Pull Request:

1. On a store, ensure that  **WooCommerce > Settings > Advanced > Features > Enable WooCommerce Analytics** is `enabled`.
2. Place several orders with a logged-in user, and with an anonymous user. Ensure no errors.
5. Navigate to **WooCommerce > Customers** and confirm that the data there is correct for the users.
    - If not correct, wait a moment and refresh until the data updates (It's the `wc-admin_import_orders` scheduled action).
6. Navigate to **WooCommerce > Order** and edit one of the new orders.
7. Confirm that the data in the Customer history metabox matches that in the Customers reports table.
8. Disable **WooCommerce Analytics** in WooCommerce Settings.
9. Confirm that the metabox no longer appears on the Order Edit page (Customer report will also no longer be available).
10. Place another order.
12. Re-enable WooCommerce Analytics.
13. Check that the Customer history is again visible on the Order Edit pages, with the correct information (it may be necessary to re-save the new order to have it imported).
14. Change one order to "Refunded" and confirm that the numbers update correctly (Total orders remains the same, Total revenue changes).
15. Change **WooCommerce > Settings > Advanced > Features > Order Attribution** to `disabled`.
16. Confirm that the Customer history metabox no longer appears on the Order Edit page (Order attribution box should also disappear).
17. Place a new order and confirm that the numbers update correctly in Customers reports.


**Bonus testing**
1. Use woocommerce/wc-smooth-generator to create 1000+ orders on a store.
2. Confirm Customer reports table numbers match Order Edit Customer history numbers.
3. Confirm that no noticeable slowdown occurs when viewing Order Edit page.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
